nifi-nar-bundles/nifi-email-bundle/nifi-email-processors/src/main/java/org/apache/nifi/processors/email/smtp/handler/SMTPMessageHandlerFactory.java

```diff
@@ -0,0 +1,147 @@
+
+            StopWatch watch = new StopWatch(false);
+
+            SMTPServer server = messageContext.getSMTPServer();
+
+            ByteArrayOutputStream baos = new ByteArrayOutputStream();
+
+            byte [] buffer = new byte[1024];
+            int rd;
+
+            while ((rd = inputStream.read(buffer, 0, buffer.length)) != -1) {
+                baos.write(buffer, 0, rd);
+            }
+            if (baos.getBufferLength() > server.getMaxMessageSize()) {
+                throw new TooMuchDataException("Data exceeds the amount allowed.");
+            }
+
+            baos.flush();
+            this.messageBody = baos.toByteArray();
+
+
+            X509Certificate[] certificates = new X509Certificate[]{};
+
+            String remoteIP = messageContext.getRemoteAddress().toString();
+            String helo = messageContext.getHelo();
+            if (messageContext.getTlsPeerCertificates() != null ){
+                certificates = (X509Certificate[]) messageContext.getTlsPeerCertificates().clone();
+            }
+
+            SmtpEvent message = new SmtpEvent(remoteIP, helo, from, recipient, certificates, messageBody);
+            try {
+                    // Try to queue the message back to the NiFi session
+                    incomingMessages.put(message);
+            } catch (InterruptedException e) {
+                // Throws an error to the  incoming server alerting about issues queuing the message in NiFi
+                logger.error("Hit an error sending message back to NiFi main thread {}. Sending and error back to SMTP client and discarding message", new Object [] {e});
+                throw new DropConnectionException(451, "NiFi something went wrong while processing you message, please retry again later");
+            }
+
+            // Once message has been sent to the queue, it should be processed by NiFi onTrigger,
+            // a flowfile created and its processed status updated before an acknowledgment is
+            // given back to the SMTP client
+            try {
+                synchronized(message) {
+                    while(! message.getProcessed()) {
+                        // Check to see if it is too late...
+                        final long serverTimeout = TimeUnit.NANOSECONDS.convert(messageContext.getSMTPServer().getConnectionTimeout(), TimeUnit.MILLISECONDS);
+                        if ( watch.getElapsed(TimeUnit.MILLISECONDS) <=  serverTimeout) {
+                            // Will should allow for a couple of retries;
+                            message.wait(messageContext.getSMTPServer().getConnectionTimeout() / 20);
+                        } else {
+                            logger.error("Did not receive the onTrigger reponse within the acceptable timeframes");
+                            throw new DropConnectionException(451, "The processing of your message timed-out, we may have received it but you better off sending it again");
```

Comment:
I don't know much about the ordering of "subethamail" aside from what you've said so I defer to you on that. That said, I think you're thinking a level higher than what I am. I am talking just about the interaction between onTrigger() and data() and the handling of the message. 

A third race condition exists, after the onTrigger polls the queue it will pull a message off the queue. This remove it from the queue and now exists as a variable in the onTrigger and data methods. In this time span the onTrigger has to write the message body to the contents, which may take a significant amount of time. While the content is being written the serverTimeout hits. When this happens it throws the error (also should remove from the queue, I commented on that) telling the user to retry. The onTrigger has no idea that the connection was dropped and will continue to process the message (because it already exists as a variable in the ontrigger). 

There may be other unwanted interactions because message exists in both but this is the one that jumped out at me.

To get around this you may be able to use the message itself as the lock in a synchronized block or add more variables to SmtpEvent to allow bidirection communication between the onTrigger and data().

Also I don't think your assumptions about how the blocking queue works is correct. If you take a look at this link[1], what makes it blocking is purely that it will let you wait for data to exist in the queue before you take from it (or wait to put data in if it is full). So if the onTrigger polls() it, it will grab whatever is the first in the queue. On the flip side, "offer" will put the data into the queue is there is space or wait until the queue does have space (with a timeout).

Lastly, I try not to make assumptions about timing (adding 5ms or waiting 20ms in a loop) because you never know what a user is going to do (maybe the timeout is only 5ms) or what system they are running on. It is much safer to not assume anything and instead just base it on things the user has configured (like the sever timeout). That way if anything goes wrong they can't blame it on the system and it was just their configuration.

[1] http://tutorials.jenkov.com/java-util-concurrent/blockingqueue.html